File: modules/meeterm-terminal/android/src/main/java/dev/meeterm/terminal/AttachmentCoreBridge.kt
package dev.meeterm.terminal

/**
 * Junction for the W2 Rust attachment contract (attachment-ffi.md plus the
 * Main amendments). Every core call passes through this object.
 *
 * Until W2's jni.rs lands on this branch the `external fun` symbols are
 * unresolved: each entry catches `UnsatisfiedLinkError` and reports
 * `unavailable` (`core_contract_pending`) so the rest of the flow degrades
 * instead of crashing. Once the .so ships the functions, the calls pass
 * straight through.
 */
internal object AttachmentCoreBridge {

  /** Attachment op result codes, mirroring the contract's error ordering. */
  private fun actionResult(code: Int, attachmentId: Long): Map<String, Any?> =
    if (code == 0) {
      AttachmentResults.accepted(attachmentId)
    } else {
      AttachmentResults.error(
        attachmentReturnCodeName(code),
        "The attachment request was rejected by the core.",
      )
    }

  private fun attachmentReturnCodeName(code: Int): String = when (code) {
    -1 -> "invalid_argument"
    -2 -> "unknown_terminal"
    -3 -> "unknown_attachment"
    -4 -> "invalid_state"
    -5 -> "source_unreadable"
    -6 -> "source_too_large"
    -7 -> "destination_not_ready"
    -8 -> "busy"
    else -> "native_error"
  }

  private fun <T> pending(action: () -> T, fallback: () -> T): T =
    try {
      action()
    } catch (_: UnsatisfiedLinkError) {
      fallback()
    }

  /**
   * `meeterm_attachment_begin`. An empty remoteDirectory selects the core
   * default `~/.local/share/meeterm/attachments`.
   * Returns the attachment id (>0), 0 on synchronous rejection, or null when
   * the core contract is not linked yet.
   */
  fun begin(
    terminalHandle: Long,
    localPath: String,
    displayName: String,
    remoteDirectory: String,
    sizeBytes: Long,
  ): Long? = pending(
    {
      MeetermNative.attachmentBegin(
        terminalHandle, localPath, displayName, remoteDirectory, sizeBytes,
      )
    },
    { null },
  )

  /** `meeterm_attachment_retry_upload`; uniform accepted/error map. */
  fun retryUpload(terminalHandle: Long, attachmentId: Long): Map<String, Any?> =
    pending(
      { actionResult(MeetermNative.attachmentRetryUpload(terminalHandle, attachmentId), attachmentId) },
      { AttachmentResults.unavailable(AttachmentLimits.REASON_CORE_PENDING) },
    )

  /** `meeterm_attachment_insert`; uniform accepted/error map. */
  fun insert(terminalHandle: Long, attachmentId: Long): Map<String, Any?> =
    pending(
      { actionResult(MeetermNative.attachmentInsert(terminalHandle, attachmentId), attachmentId) },
      { AttachmentResults.unavailable(AttachmentLimits.REASON_CORE_PENDING) },
    )

  /** `meeterm_attachment_cancel`; uniform accepted/error map. */
  fun cancel(attachmentId: Long): Map<String, Any?> =
    pending(
      { actionResult(MeetermNative.attachmentCancel(attachmentId), attachmentId) },
      { AttachmentResults.unavailable(AttachmentLimits.REASON_CORE_PENDING) },
    )

  /** `meeterm_attachment_dispose`; uniform accepted/error map. */
  fun dispose(attachmentId: Long): Map<String, Any?> =
    pending(
      { actionResult(MeetermNative.attachmentDispose(attachmentId), attachmentId) },
      { AttachmentResults.unavailable(AttachmentLimits.REASON_CORE_PENDING) },
    )

  /** `meeterm_attachment_delete_remote`; uniform accepted/error map. */
  fun deleteRemote(terminalHandle: Long, attachmentId: Long): Map<String, Any?> =
    pending(
      { actionResult(MeetermNative.attachmentDeleteRemote(terminalHandle, attachmentId), attachmentId) },
      { AttachmentResults.unavailable(AttachmentLimits.REASON_CORE_PENDING) },
    )

  /** `meeterm_attachment_snapshot`; decoded record, null when unknown. */
  fun snapshot(attachmentId: Long): AttachmentOperation? =
    pending(
      { MeetermNative.attachmentSnapshot(attachmentId)?.let(AttachmentOperationCodec::decode) },
      { null },
    )
}

File: modules/meeterm-terminal/ios/AttachmentCoreBridge.swift
import Foundation

/**
 * Exclusive seam between the adapter and the Rust attachment ABI
 * (attachment-ffi.md + Main amendments). Every core call passes through this
 * enum, which maps one-to-one onto `MeetermCore.attachment*` — no other Swift
 * file may reference the C symbols. A negative contract code is surfaced with
 * its sanitized error name.
 */
enum AttachmentCoreBridge {
  private static func actionResult(_ code: Int32, attachmentId: UInt64) -> [String: Any] {
    code == 0
      ? AttachmentResults.accepted(attachmentId)
      : AttachmentResults.error(
          AttachmentResults.returnCodeError(code),
          "The attachment request was rejected by the core."
        )
  }

  /**
   * `meeterm_attachment_intent`: record the destination intent for the
   * picked pane's native terminal. The core resolves the owning SSH
   * connection itself — the adapter never passes an owner id.
   * Returns the opaque intent id (>0) or 0 when the pane is not a usable
   * destination.
   */
  static func intent(targetTerminalId: UInt64) -> UInt64 {
    MeetermCore.attachmentIntent(targetTerminalId: targetTerminalId)
  }

  /// `meeterm_attachment_intent_dispose`; idempotent.
  static func intentDispose(intentId: UInt64) {
    _ = MeetermCore.attachmentIntentDispose(intentId: intentId)
  }

  /// `meeterm_attachment_begin`. An empty remoteDirectory selects the core
  /// default `~/.local/share/meeterm/attachments`. Returns the attachment id
  /// (>0) or nil on synchronous rejection (unknown intent, unreadable file,
  /// changed/missing destination, or a second live op).
  static func begin(
    intentId: UInt64,
    localPath: String,
    displayName: String,
    remoteDirectory: String,
    sizeBytes: UInt64
  ) -> UInt64? {
    let attachmentId = MeetermCore.attachmentBegin(
      intentId: intentId,
      localPath: localPath,
      displayName: displayName,
      remoteDirectory: remoteDirectory,
      sizeBytes: sizeBytes
    )
    return attachmentId > 0 ? attachmentId : nil
  }

  /// `meeterm_attachment_retry_upload`; uniform accepted/error map.
  /// `targetTerminalId` must be the intent's recorded pane terminal.
  static func retryUpload(terminalId: UInt64, attachmentId: UInt64) -> [String: Any] {
    actionResult(
      MeetermCore.attachmentRetryUpload(targetTerminalId: terminalId, attachmentId: attachmentId),
      attachmentId: attachmentId
    )
  }

  /// `meeterm_attachment_insert`; uniform accepted/error map.
  static func insert(terminalId: UInt64, attachmentId: UInt64) -> [String: Any] {
    actionResult(
      MeetermCore.attachmentInsert(targetTerminalId: terminalId, attachmentId: attachmentId),
      attachmentId: attachmentId
    )
  }

  /// `meeterm_attachment_cancel`; uniform accepted/error map.
  static func cancel(attachmentId: UInt64) -> [String: Any] {
    actionResult(MeetermCore.attachmentCancel(attachmentId: attachmentId), attachmentId: attachmentId)
  }

  /// `meeterm_attachment_dispose`; uniform accepted/error map.
  static func dispose(attachmentId: UInt64) -> [String: Any] {
    actionResult(MeetermCore.attachmentDispose(attachmentId: attachmentId), attachmentId: attachmentId)
  }

  /// `meeterm_attachment_delete_remote`; uniform accepted/error map.
  /// `targetTerminalId` must be the intent's recorded pane terminal.
  static func deleteRemote(terminalId: UInt64, attachmentId: UInt64) -> [String: Any] {
    actionResult(
      MeetermCore.attachmentDeleteRemote(targetTerminalId: terminalId, attachmentId: attachmentId),
      attachmentId: attachmentId
    )
  }

  /// `meeterm_attachment_snapshot` decoded by the pure codec.
  static func snapshot(attachmentId: UInt64) -> AttachmentOperation? {
    guard let bytes = MeetermCore.attachmentSnapshot(attachmentId: attachmentId) else {
      return nil
    }
    return AttachmentOperationCodec.decode(bytes)
  }
}
